x/bank/types/balance.go

```diff
@@ -0,0 +1,93 @@
+package types
+
+import (
+	"bytes"
+	"encoding/json"
+	fmt "fmt"
+	"sort"
+
+	"github.com/cosmos/cosmos-sdk/codec"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/cosmos/cosmos-sdk/x/bank/exported"
+)
+
+var _ exported.GenesisBalance = (*Balance)(nil)
+
+// GetAddress returns the account address of the Balance object.
+func (b Balance) GetAddress() sdk.AccAddress {
+	addr, _ := sdk.AccAddressFromBech32(b.Address)
+	return addr
+}
+
+// GetCoins returns the account coins of the Balance object.
+func (b Balance) GetCoins() sdk.Coins {
+	return b.Coins
+}
+
+// Validate checks for address and coins correctness.
+func (b Balance) Validate() error {
+	_, err := sdk.AccAddressFromBech32(b.Address)
+	if err != nil {
+		return err
+	}
+
+	if len(b.Coins) == 0 {
+		return fmt.Errorf("empty or nil coins for address %s", b.Address)
+	}
+
+	seenDenoms := make(map[string]bool)
+
+	// NOTE: we perform a custom validation since the coins.Validate function
+	// errors on zero balance coins
+	for _, coin := range b.Coins {
+		if seenDenoms[coin.Denom] {
+			return fmt.Errorf("duplicate denomination %s", coin.Denom)
+		}
+
+		if err := sdk.ValidateDenom(coin.Denom); err != nil {
+			return err
+		}
+
+		if coin.IsNegative() {
+			return fmt.Errorf("coin %s amount is cannot be negative", coin.Denom)
+		}
```

Comment:
We should put the cheaper call before to avoid wasted CPU cycles and from look at the code flow, most likely coins will be negative than to be within this code path and have invalid denominations.